src/bootstrap/native.rs

```diff
@@ -111,6 +115,201 @@ pub fn prebuilt_llvm_config(
+        println!("help: maybe your repository history is too shallow?");
+        println!("help: consider disabling `download-ci-llvm`");
+        println!("help: or fetch enough history to include one upstream commit");
+        panic!();
+    }
+
+    let llvm_stamp = builder.config.ci_llvm_root().join(".llvm-stamp");
+    let key = format!("{}{}", llvm_sha, config.llvm_assertions);
+    if program_out_of_date(&llvm_stamp, &key) {
+        download_ci_llvm(builder, &llvm_sha);
+        // with output(self.llvm_stamp()) as llvm_stamp:
+        //     llvm_stamp.write(llvm_sha + str(llvm_assertions))
+        if !config.dry_run {
+            t!(fs::write(llvm_stamp, key));
+        }
+    }
+
+    // if self.program_out_of_date(self.llvm_stamp(), llvm_sha + str(llvm_assertions)):
+    //     self._download_ci_llvm(llvm_sha, llvm_assertions)
+    //     for binary in ["llvm-config", "FileCheck"]:
+    //         self.fix_bin_or_dylib(os.path.join(llvm_root, "bin", binary))
+    //     for lib in os.listdir(llvm_lib):
+    //         if lib.endswith(".so"):
+    //             self.fix_bin_or_dylib(os.path.join(llvm_lib, lib))
+    //     with output(self.llvm_stamp()) as llvm_stamp:
+    //         llvm_stamp.write(llvm_sha + str(llvm_assertions))
+}
+
+fn download_ci_llvm(builder: &Builder<'_>, llvm_sha: &str) {
+    if builder.config.dry_run {
+        return;
+    }
+
+    let llvm_assertions = builder.config.llvm_assertions;
+
+    let cache_prefix = format!("llvm-{}-{}", llvm_sha, llvm_assertions);
+    let cache_dst = builder.out.join("cache");
+    let rustc_cache = cache_dst.join(cache_prefix);
+    if !rustc_cache.exists() {
+        t!(fs::create_dir_all(&rustc_cache));
+    }
+    let base = "https://ci-artifacts.rust-lang.org";
+    let url = if llvm_assertions {
+        format!("rustc-builds-alt/{}", llvm_sha)
+    } else {
+        format!("rustc-builds/{}", llvm_sha)
+    };
+    let filename = String::from("rust-dev-nightly-") + &builder.build.build.triple + ".tar.xz";
+    let tarball = rustc_cache.join(&filename);
+    if !tarball.exists() {
+        download_component(builder, base, &format!("{}/{}", url, filename), &tarball);
+        // get(
+        //     base,
+        //     "{}/{}".format(url, filename),
+        //     tarball,
+        //     self.checksums_sha256,
+        //     verbose=self.verbose,
+        //     do_verify=False,
+        //     help_on_error=help_on_error,
+        // )
+    }
+    let llvm_root = builder.config.ci_llvm_root();
+    unpack(builder, &tarball, &llvm_root);
+
+    // unpack(tarball, tarball_suffix, self.llvm_root(),
+    //  match="rust-dev",
+    //  verbose=self.verbose)
+
+    // TODO: `fix_bin_or_dylib`
+    // let llvm_lib = builder.llvm_root().join("lib");
+}
+
+fn download_component(builder: &Builder<'_>, base: &str, url: &str, dest_path: &Path) {
+    let tempfile = t!(tempfile::NamedTempFile::new());
```

Comment:
bootstrap.py downloads to `build/cache`, right? That would allow avoiding an extra rust dependency for rustbuild.

Reply:
No, it downloads to a temporary directory, then moves the file to `build/cache`. That's necessary to avoid having a partially downloaded file in cache, since bootstrap doesn't verify checksums before unpacking, only just after downloading, so otherwise it wouldn't be resistant to a crash halfway through the download.
https://github.com/rust-lang/rust/blob/beb480d908f091e7c42c866e13641f26b2ee230b/src/bootstrap/bootstrap.py#L67

Reply:
Hmm, I suppose we could download to `build/cache/temp` or something like that - I'd like to land this first, but it could be done in a follow-up. It does mean if bootstrap gets interrupted while running, we'll leave around temp files forever, but that doesn't seem like the *worst* thing in the world since `build` is already assumed to be temp files.

Reply:
Bootstrap could remove partially downloaded files from a previous invocation.

Reply:
I think build/cache/temp would be a reasonable solution -- I'd prefer to avoid the tempfile dependency and cross-device detection in rename() that is otherwise necessary.